File: 1004-max-consecutive-ones-iii/1004. Max Consecutive Ones III.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        left = 0
        zero_count = 0
        max_length = 0

        for right in range(len(nums)):

            # Add the new element
            if nums[right] == 0:
                zero_count += 1

            # Too many zeros → shrink window
            while zero_count > k:
                if nums[left] == 0:
                    zero_count -= 1
                left += 1

            # Current window is valid
            max_length = max(max_length, right - left + 1)

        return max_length

File: 1004-max-consecutive-ones-iii/1004. Max Consecutive Ones III.java
class Solution {
    public int longestOnes(int[] nums, int k) {

        int i = 0, j = 0;
        int n = nums.length;
        int curr = 0;
        int ans = 0;
        int last = 0;
        ArrayList<Integer> zeros = new ArrayList<>();

        for (int z = 0; z < n; z++)
            if (nums[z] == 0)
                zeros.add(z);

        while (j < n) {
            if (nums[j] != 0) {
                ans = Math.max(ans, (j - i + 1));;
                j++;
                continue;
            }

            if (nums[j] == 0) {

                if (curr == k) {
                    // ans = Math.max(ans, (j - i + 1));
                    i = zeros.get(last) + 1;
                    curr--;
                    last++;
                }

                curr++;

                if(curr <= k)ans = Math.max(ans, (j - i + 1));
                j++;
            }
        }

        return ans;

    }
}